tt_eager/tt_dnn/op_library/bmm/multi_core_reuse_mcast_2d_optimized/bmm_op_multi_core_reuse_mcast_2d_optimized.cpp

```diff
@@ -113,77 +112,145 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
 
     CoreRange top_row_except_corner(
         {(std::size_t)start_core_x + 1, (std::size_t)start_core_y},
-        {(std::size_t)start_core_x + num_cores_c - 1, (std::size_t)start_core_y});
+        {(std::size_t)start_core_x + num_cores_with_work_c - 1, (std::size_t)start_core_y});
 
+    ////////////////////////////////////////////////////////////////////////////
+    //                      INTERLEAVED SENDER/RECEIVER
+    ////////////////////////////////////////////////////////////////////////////
     CoreRange in0_sender = left_column;
     CoreRange in0_sender_in1_receiver = left_column_except_corner;
     CoreRange in1_sender = top_row;
     CoreRange in0_receiver_in1_sender = top_row_except_corner;
-
-    uint32_t in0_end = num_cores_r - 1;
-    uint32_t in1_end = num_cores_c - 1;
-
-    // in1 is the reader of weights/output writer, and we choose to make it use the optimized reader noc
-    tt_metal::NOC in0_noc = detail::GetPreferredNOCForDRAMWrite(device->arch());
-    tt_metal::NOC in1_noc = detail::GetPreferredNOCForDRAMRead(device->arch());
-    tt_metal::NOC in0_split_noc = detail::GetPreferredNOCForDRAMRead(device->arch());
-    tt_metal::NOC in1_split_noc = detail::GetPreferredNOCForDRAMWrite(device->arch());
     if (transpose_mcast) {
         std::swap(in0_sender, in1_sender);
         std::swap(in0_sender_in1_receiver, in0_receiver_in1_sender);
-        std::swap(in0_end, in1_end);
     }
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                      IN0 SHARDED SENDER/RECEIVER
+    ////////////////////////////////////////////////////////////////////////////
+    uint32_t num_cores_c = num_cores_with_work_c;
+    uint32_t num_cores_r = num_cores_with_work_r;
+    uint32_t in0_mcast_receiver_grid_diff_coord_start;
+    uint32_t in0_mcast_receiver_grid_diff_coord_end;
+    std::vector<uint32_t> in0_mcast_noc_x;
+    std::vector<uint32_t> in0_mcast_noc_y;
+    uint32_t in0_sender_num_cores_along_width;
+
+    // Only used for in0 block sharded
+    CoreRange in0_mcast_cores_with_work_and_in_receiver_grid(
+        {(std::size_t)start_core_x, (std::size_t)start_core_y},
+        {(std::size_t)start_core_x, (std::size_t)start_core_y});  // Placeholder
+    CoreRange in0_mcast_cores_without_work_and_not_in_receiver_grid(
+        {(std::size_t)start_core_x, (std::size_t)start_core_y},
+        {(std::size_t)start_core_x, (std::size_t)start_core_y});  // Placeholder
     if (in0_block_sharded) {
-        in0_sender = all_cores;
+        in0_mcast_cores_with_work_and_in_receiver_grid = all_cores_with_work;
+
+        CoreCoord in0_shard_grid = in0_buffer->shard_spec().grid().bounding_box().grid_size();
+        // in0 shard grid already accounts for transpose_mcast
+        // ie. If transpose_mcast, in0 width is along y
+        in0_sender_num_cores_along_width = transpose_mcast ? in0_shard_grid.y : in0_shard_grid.x;
+        if (in0_sender_num_cores_along_width > num_blocks_x) {
```

Comment:
Under what circumstances is `in0_sender_num_cores_along_width <= num_blocks_x` and what happens further in the code?

Reply:
It happens when in0 shard width is less than the number of cores along output width. The code already works for this case since we used to mcast to the full output grid anyways. But now, it's cleaner and more explicit

Reply:
Ok. Thanks.